### PR TITLE
Fix LastWriteTime and LastAccessTime of a symlink on Windows and Unix

### LINKED ISSUE #38824: Cannot change LastWriteTime or LastAccessTime of a symlink
### Description
It does not appear to be possible to change the LastWriteTime or LastAccessTime of a symlink. It works on hardlinks.

```
PS C:\src\t> $SleepTime = 61
PS C:\src\t> $BaseDir = 'C:\src\t'
PS C:\src\t> $BaseFile = Join-Path -Path $BaseDir -ChildPath 'yy.txt'
PS C:\src\t> $SymlinkFile = Join-Path -Path $BaseDir -ChildPath 'yy-slink.txt'
PS C:\src\t> $HardlinkFile = Join-Path -Path $BaseDir -ChildPath 'yy-hlink.txt'
PS C:\src\t> if (Test-Path -Path $BaseFile) { Remove-Item -Path $BaseFile }
PS C:\src\t> if (Test-Path -Path $SymlinkFile) { Remove-Item -Path $SymlinkFile }
PS C:\src\t> if (Test-Path -Path $HardlinkFile) { Remove-Item -Path $HardlinkFile }
PS C:\src\t> New-Item -Path $BaseFile -ItemType File -Value "this is a new file`r`n"


    Directory: C:\src\t

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---          2020-07-06    14:23             20 yy.txt

PS C:\src\t> New-Item -ItemType SymbolicLink -Path $SymlinkFile -Target $BaseFile


    Directory: C:\src\t

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
la---          2020-07-06    14:23              0 yy-slink.txt -> C:\src\t\yy.txt

PS C:\src\t> Get-ChildItem -Path 'C:\src\t' -Filter 'yy*.txt' | Format-List 'Name','*Time'

Name           : yy.txt
CreationTime   : 2020-07-06 14:23:41
LastAccessTime : 2020-07-06 14:23:41
LastWriteTime  : 2020-07-06 14:23:41

Name           : yy-slink.txt
CreationTime   : 2020-07-06 14:23:41
LastAccessTime : 2020-07-06 14:23:41
LastWriteTime  : 2020-07-06 14:23:41


PS C:\src\t> Start-Sleep -Seconds $SleepTime
PS C:\src\t> $NewTimestamp = Get-Date
PS C:\src\t> Set-ItemProperty -Path $SymlinkFile -Name LastAccessTime -Value $NewTimestamp
PS C:\src\t> Set-ItemProperty -Path $SymlinkFile -Name LastWriteTime -Value $NewTimestamp
PS C:\src\t> Start-Sleep -Seconds 5
PS C:\src\t> Get-ChildItem -Path $BaseDir -Filter 'yy*.txt' | Format-List 'Name','*Time'

Name           : yy.txt
CreationTime   : 2020-07-06 14:23:41
LastAccessTime : 2020-07-06 14:24:42
LastWriteTime  : 2020-07-06 14:24:42

Name           : yy-slink.txt
CreationTime   : 2020-07-06 14:23:41
LastAccessTime : 2020-07-06 14:23:41
LastWriteTime  : 2020-07-06 14:23:41


PS C:\src\t> & cmd.exe /C "DIR /TA yy*"
 Volume in drive C has no label.
 Volume Serial Number is 0E33-300C

 Directory of C:\src\t

2020-07-06  14:23    <SYMLINK>      yy-slink.txt [C:\src\t\yy.txt]
2020-07-06  14:24                20 yy.txt
               2 File(s)             20 bytes
               0 Dir(s)  739,939,135,488 bytes free
PS C:\src\t> & cmd.exe /C "DIR /TW yy*"
 Volume in drive C has no label.
 Volume Serial Number is 0E33-300C

 Directory of C:\src\t

2020-07-06  14:23    <SYMLINK>      yy-slink.txt [C:\src\t\yy.txt]
2020-07-06  14:24                20 yy.txt
               2 File(s)             20 bytes
               0 Dir(s)  739,939,135,488 bytes free
PS C:\src\t> New-Item -ItemType HardLink -Path $HardlinkFile -Target $BaseFile


    Directory: C:\src\t

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---          2020-07-06    14:24             20 yy-hlink.txt

PS C:\src\t> Get-ChildItem -Path $BaseDir -Filter 'yy*.txt' | Format-List 'Name','*Time'

Name           : yy.txt
CreationTime   : 2020-07-06 14:23:41
LastAccessTime : 2020-07-06 14:24:42
LastWriteTime  : 2020-07-06 14:24:42

Name           : yy-hlink.txt
CreationTime   : 2020-07-06 14:23:41
LastAccessTime : 2020-07-06 14:24:42
LastWriteTime  : 2020-07-06 14:24:42

Name           : yy-slink.txt
CreationTime   : 2020-07-06 14:23:41
LastAccessTime : 2020-07-06 14:23:41
LastWriteTime  : 2020-07-06 14:23:41


PS C:\src\t> Start-Sleep -Seconds $SleepTime
PS C:\src\t> $NewTimestamp = Get-Date
PS C:\src\t> Set-ItemProperty -Path $HardlinkFile -Name LastAccessTime -Value $NewTimestamp
PS C:\src\t> Set-ItemProperty -Path $HardlinkFile -Name LastWriteTime -Value $NewTimestamp
PS C:\src\t> Start-Sleep -Seconds 5
PS C:\src\t> Get-ChildItem -Path $BaseDir -Filter 'yy*.txt' | Format-List 'Name','*Time'

Name           : yy.txt
CreationTime   : 2020-07-06 14:23:41
LastAccessTime : 2020-07-06 14:24:42
LastWriteTime  : 2020-07-06 14:24:42

Name           : yy-hlink.txt
CreationTime   : 2020-07-06 14:23:41
LastAccessTime : 2020-07-06 14:25:50
LastWriteTime  : 2020-07-06 14:25:50

Name           : yy-slink.txt
CreationTime   : 2020-07-06 14:23:41
LastAccessTime : 2020-07-06 14:23:41
LastWriteTime  : 2020-07-06 14:23:41


PS C:\src\t> & cmd.exe /C "DIR /TA yy*"
 Volume in drive C has no label.
 Volume Serial Number is 0E33-300C

 Directory of C:\src\t

2020-07-06  14:25                20 yy-hlink.txt
2020-07-06  14:23    <SYMLINK>      yy-slink.txt [C:\src\t\yy.txt]
2020-07-06  14:24                20 yy.txt
               3 File(s)             40 bytes
               0 Dir(s)  739,939,672,064 bytes free
PS C:\src\t> & cmd.exe /C "DIR /TW yy*"
 Volume in drive C has no label.
 Volume Serial Number is 0E33-300C

 Directory of C:\src\t

2020-07-06  14:25                20 yy-hlink.txt
2020-07-06  14:23    <SYMLINK>      yy-slink.txt [C:\src\t\yy.txt]
2020-07-06  14:24                20 yy.txt
               3 File(s)             40 bytes
               0 Dir(s)  739,939,672,064 bytes free

```
### Configuration
```
C:>dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.100-preview.5.20279.10
 Commit:    8139f1b74e

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Users\pwatson\AppData\Local\Microsoft\dotnet\sdk\5.0.100-preview.5.20279.10\

Host (useful for support):
  Version: 5.0.0-preview.5.20278.1
  Commit:  4ae4e2fe08

.NET SDKs installed:
  2.2.402 [C:\Program Files\dotnet\sdk]
  3.0.100-rc1-014190 [C:\Program Files\dotnet\sdk]
  3.1.301 [C:\Program Files\dotnet\sdk]
  5.0.100-preview.5.20279.10 [C:\Users\pwatson\AppData\Local\Microsoft\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.19 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.19 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-rc1.19457.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0-preview.5.20279.2 [C:\Users\pwatson\AppData\Local\Microsoft\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.19 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-rc1-19456-20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0-preview.5.20278.1 [C:\Users\pwatson\AppData\Local\Microsoft\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0-rc1-19456-20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0-preview.5.20278.3 [C:\Users\pwatson\AppData\Local\Microsoft\dotnet\shared\Microsoft.WindowsDesktop.App]
```
### Regression?
Unknown
### Other information
```
PS C:\src\t> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.0.1
PSEdition                      Core
GitCommitId                    7.0.1
OS                             Microsoft Windows 10.0.18363
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by hamarb123**
Fixes #38824.
Fixes setting the LastWriteTime and the LastAccessTime on a symlink on Windows and Unix.
Somewhat reliant on https://github.com/dotnet/runtime/pull/49555.
Reverts the Windows changes of the commit https://github.com/dotnet/runtime/commit/a617a01195b4dad3cb5f300962ad843b46d4f175 (on the https://github.com/dotnet/runtime/pull/49555 PR) to fix #38824 and for Unix: https://github.com/hamarb123/runtime/commit/f84293583bb0c8e1060bdacfb80158a6c64c5fbd.

### CODE REVIEW

**Review by carlossanlop — COMMENTED** [review]
The change looks good so far, @hamarb123, thanks for the fix.

As mentioned in your description, unit tests should be included, so please ping me when you add them.

**Comment by jeffhandley** [comment]
Hi @hamarb123 -- it looks like the we were waiting for some unit tests to be added. Let us know if you need any help or guidance on that. Thanks!

**Comment by hamarb123** [comment]
@jeffhandley, unit tests are in #49555, they just need to be enabled on Windows when #49555 is merged.

**Comment by jeffhandley** [comment]
Ah; I'm sorry I overlooked that @hamarb123!

**Comment by jeffhandley** [comment]
@Jozkee This PR is assigned to you for follow-up/decision before the RC1 snap.

**Comment by hamarb123** [comment]
I'm separating out the symlink code into this PR for Unix from #49555 as per https://github.com/dotnet/runtime/pull/49555#issuecomment-904413721.

**Comment by hamarb123** [comment]
I'm not sure how to make this PR include the changes of the other PR without having it as seperate commits that need to also be merged, so for now I'll just remove the relevant code from #49555 and add it here once it's merged. If it is possible, please let me know how I can do it because this PR needs to change files added in #49555.

**Comment by danmoseley** [comment]
I see this is blocked on https://github.com/dotnet/runtime/pull/49555. Perhaps we/you can reopen this PR when this is unblocked.

**Comment by hamarb123** [comment]
Hi, please reopen this PR. Thanks!

**Comment by jeffhandley** [comment]
We're inconsistent with how we use the assignees field and I should have commented what my intent was. My intent is for you (as the author) to drive this PR through to completion but to have @Jozkee assigned as the team member accountable for ensuring we're supporting you as needed. You're still assigned the issue though. Is that OK?

(We've been discussing some better practices for ensuring we're not leaving contributors hanging for long periods of time).

**Review by jozkee — COMMENTED** [review]
@hamarb123 can you please address using `lutimes` instead of `utimes` as pointed out here https://github.com/dotnet/runtime/pull/52639/files#r737921487?

**Review by jozkee — APPROVED** [review]
Otherwise, LGTM. Thanks!

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (Jozkee) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:112]: > It would be good to add some randomization here. One setter could cause subsequent getters to succeed even if their setter is broken.

Are you referring to randomize the date that is set in `SettingUpdatesPropertiesCore`, or randomize the dates of the target, or both?

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (Jozkee) [inline src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:116]: This will be `null` when `T` is `string`. To avoid that you can add an abstract method `GetFileSystemInfo(T item)` and define how to get a FileSystemInfo for each derived class.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (Jozkee) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:112]:
```suggestion
            if (!targetExists) 
            {
                SettingUpdatesPropertiesCore(item);
            }
```

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:117]: I'm not familiar with when `atime` updates. When a symlink is created on Unix, it checks if the target exists and then if it's a file or a directory. To me this sounds like accessing, please correct me if there's no chance of `atime` updating. @Jozkee, do you know if `atime` might update?

File: src/libraries/System.IO.FileSystem/tests/Directory/GetSetTimes.cs
Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Directory/GetSetTimes.cs:14]: In the next commit.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:120]: In the next commit.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:111]: In the next commit.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:112]: This is in the next commit.

Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:112]: > Are you referring to randomize the date that is set in `SettingUpdatesPropertiesCore`, or randomize the dates of the target, or both?

I think they're referring to randomising both, but it's unnecessary since you said path names are randomised.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:96]: I thought it might be a good test case and @Jozkee agreed, it is probably unnecessary though. https://github.com/dotnet/runtime/pull/52639#discussion_r736013152

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:116]: Hi, the point of this code is to call `FileSystemInfo.Refresh` to invalidate the times cache before getting the times. If it uses `File` or `Directory`, it will not be cached (right?). Hence if creating a link can change these times, the cache needs to be refreshed. Also, it definitely needs to be refreshed after running the test on the link since that's the point of the extra check.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (tmds) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:95]: @Jozkee the additional argument checks whether both relative and full paths work as the symlink target path. It's not related to the functionality that gets tested here.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (tmds) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:119]: `atime` changes when a file is opened for reading. That does not happen here.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (tmds) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs]: > Are you referring to randomize the date that is set in SettingUpdatesPropertiesCore, or randomize the dates of the target, or both?

`SettingUpdatesPropertiesCore`. The getters/setters run on the same path, so the path being unique doesn't detect issues where one setter affects multiple getters.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:119]: 👍 ok, I'll remove the initial refresh. If the assert fails (which it shouldn't), I'll put it back.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs]: > `SettingUpdatesPropertiesCore`. The getters/setters run on the same path, so the path being unique doesn't detect issues where one setter affects multiple getters.

There's a different test for that (for normal files). It's assumed that it will also work on symlinks if times are set correctly on symlinks rather than the targets (which this test checks), since there is no code that currently, and shouldn't ever probably be any time setting/getting code that treats symlinks differently to normal files (it would likely fail this test also if it did anyway). Note the condition which is most likely for one setter affecting multiple getters is "when the modification time is less than the creation time" (if code was done from scratch for macOS without reading some of the comments would be when this is an issue), so simply testing if particular setters change more than 1 getter isn't good enough.
I do think it's probably unnecessary, but if @Jozkee wants, I could also implement the `SettingUpdatesPropertiesAfterAnother` test for symlinks, but again, I think it's probably overkill since it's being checked for normal files.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:95]: The reason that we included it here is we're checking if it changes the date on the target file as well, so all variants of the 'target' should be included IMO (which is what it currently is: relative/absolute, exists/doesn't, file/folder, info/static).

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:118]: @tmds says it will work without the first one, so I'm only keeping the 2nd one (which definitely is necessary, else it'll return the cached times). https://github.com/dotnet/runtime/pull/52639#discussion_r739761174

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (tmds) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:95]: > all variants of the 'target' should be included IMO

Another variant would be using multiple links to get to the target.

Because the target gets resolved by the OS, I don't see the need use different target paths.

I'm ok if you want to keep it.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (tmds) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs]: My comment is not specific to this PR. It applies also to the `SettingUpdatesProperties` test as it was.

Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs]: See the `SettingUpdatesPropertiesAfterAnother` test.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:95]: > Another variant would be using multiple links to get to the target.

That's true. I don't see it as necessary personally.

> Because the target gets resolved by the OS, I don't see the need use different target paths.

Probably fair. I'd rather it be there, if @Jozkee wants I can remove it.

> I'm ok if you want to keep it.

👍 thanks

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (tmds) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs]: Ah, yes, `SettingUpdatesPropertiesAfterAnother` will catch it.

**Review by tmds — APPROVED** [review]
lgtm, thanks!

File: src/libraries/Native/Unix/System.Native/pal_time.c
Comment (Jozkee) [inline src/libraries/Native/Unix/System.Native/pal_time.c:44]: You can re-use this code for lutimes and utimes i.e: the `#elif` can be shrinked to lutimes vs utimes.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (Jozkee) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:95]: What @tmds says makes sense. You can remove `targetIsRelative` if you want.

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (Jozkee) [inline src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs:124]:
```suggestion
                if (target is FileSystemInfo fsi)
                { 
                    fsi.Refresh();
                }
```

File: src/libraries/Native/Unix/System.Native/pal_time.c
Comment (hamarb123) [reply src/libraries/Native/Unix/System.Native/pal_time.c]: I've combined. Let me know if you want it combined differently (ie. #if ing a whole line rather than a function name).

File: src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs
Comment (hamarb123) [reply src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetTimes.cs]: Done

**Review by jozkee — APPROVED** [review]

**Comment by Jozkee** [comment]
Breaking change doc issue: https://github.com/dotnet/docs/issues/31483

**Pull request merged**
